src/libraries/System.Private.CoreLib/src/System/Reflection/InvokerEmitUtil.cs

```diff
@@ -0,0 +1,331 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection.Emit;
+using System.Runtime.CompilerServices;
+
+namespace System.Reflection
+{
+    internal static class InvokerEmitUtil
+    {
+        // This is an instance method to avoid overhead of shuffle thunk.
+        internal unsafe delegate object? InvokeFunc<T>(T obj, object? target, IntPtr* arguments);
+
+        // Avoid high arg count due to increased stack.
+        // This also allows the local variables to use that more smaller encoded _S opcode variants.
+        public const int MaxArgumentCount = 64;
+
+        public static unsafe InvokeFunc<T> CreateInvokeDelegate<T>(MethodBase method)
+        {
+            Debug.Assert(!method.ContainsGenericParameters);
+
+            ParameterInfo[] parameters = method.GetParametersNoCopy();
+            Debug.Assert(parameters.Length <= MaxArgumentCount);
+
+            Type[] delegateParameters = new Type[3] { typeof(T), typeof(object), typeof(IntPtr*) };
+
+            // We could use the overload with 'owner' in order to associate to this module.
+            var dm = new DynamicMethod(
+                "InvokeStub_" + method.DeclaringType!.Name + "." + method.Name,
+                returnType: typeof(object),
+                delegateParameters,
+                owner: typeof(T),
+                skipVisibility: true);
+
+            ILGenerator il = dm.GetILGenerator();
+
+            if (parameters.Length == 0)
+            {
+                HandleThisPointer(il, method);
+                Invoke(il, method);
+            }
+            else
+            {
+                HandleThisPointer(il, method);
+                PushArguments(il, parameters, out NullableRefInfo[]? byRefNullables, out int byRefNullableCount);
+                Invoke(il, method);
+
+                if (byRefNullableCount > 0)
+                {
+                    Debug.Assert(byRefNullables != null); ;
+                    UpdateNullables(il, parameters, byRefNullables, byRefNullableCount);
+                }
+            }
+
+            HandleReturn(il, method);
+
+            return (InvokeFunc<T>)dm.CreateDelegate(typeof(InvokeFunc<T>));
+        }
+
+        private static void HandleThisPointer(ILGenerator il, MethodBase method)
+        {
+            bool emitNew = method is RuntimeConstructorInfo;
+            bool hasThis = !(emitNew || method.IsStatic);
+
+            if (hasThis)
+            {
+                il.Emit(OpCodes.Ldarg_1);
+                if (method.DeclaringType!.IsValueType)
+                {
+                    il.Emit(OpCodes.Unbox, method.DeclaringType);
```

Comment:
The boxing/unboxing of `this` and return value should be moved to the static code too.

Reply:
Yes per discussion the forcing factor will be supporting the new byref-like type APIs possibly using `TypedReference`. That will enable true by-ref returns (with aliasing) and ability to invoke a target value type (and eventually byref-like types) without boxing or making a copy of the target.